Document/0x04g-Testing-Cryptography.md

```diff
@@ -69,77 +89,93 @@ See also the following best practice documents for recommendations:
 
 ##### Weak Key Generation Functions
 
-Cryptographic algorithms, such as symmetric encryption or MACs, expect a secret input of a given size, e.g. 128 or 256 bit. A native implementation might use the user-supplied password directly as an input key. There are a couple of problems with this approach:
+Cryptographic algorithms (such as symmetric encryption or MACs) expect a secret input of a given size (for example, 128 or 256 bit). A native implementation might use the user-supplied password directly as an input key. Using a user-supplied password as an input key has the following problems:
 
-- If the password is smaller than the key, then not the full key-space is used (the rest is padded, sometimes even with spaces)
-- A user-supplied password will realistically consist mostly of displayable and pronounceable characters. So instead of the full entropy, i.e. 2<sup>8</sup> when using ASCII, only a small subset is used (approx. 2<sup>6</sup>).
-- If two users select the same password an attacker can match the encrypted files. This opens up the possibility of rainbow table attacks.
+- If the password is smaller than the key, the full key space isn't used. The remaining space is padded (spaces are sometimes used for padding).
+- A user-supplied password will realistically consist mostly of displayable and pronounceable characters. Therefore, only a subset of all possible ASCII characters are used and entropy is decresed (from 2<sup>8</sup> to approximately 2<sup>6</sup>).
+- If two users select the same password, an attacker can match the encrypted files. This makes rainbow table attacks a possibility.
 
-Verify that no password is directly passed into an encryption function. Instead, the user-supplied password should be passed into a salted hash function or KDF to create the cryptographic key. Choose and appropriate iteration count when using password derivation functions. For example, [NIST recommends and iteration count of at least 10,000 for PBKDF2](https://pages.nist.gov/800-63-3/sp800-63b.html#sec5 "NIST Special Publication 800-63B").
+> How can the password be smaller then the key when the password is being used as the key? Maybe I'm just not understanding, or there is some inconsistency with the lead-in paragraph and the first bullet?
+
+Ensure that passwords arn't directly passed into an encryption function. Instead, the user-supplied password should be passed into a salted hash function or KDF to create a cryptographic key. Choose an appropriate iteration count when using password derivation functions. For example, [NIST recommends and iteration count of at least 10,000 for PBKDF2](https://pages.nist.gov/800-63-3/sp800-63b.html#sec5 "NIST Special Publication 800-63B").
```

Comment:
> arn't

aren't

> into a salted hash function or KDF

I think we can recommend a key derivation function (KDF) to derive keys from passwords, and remove te "salted hash".

Reply:
Sorry to make you spend time commenting on typos, that wasn't my intent. This was just a really rough first pass without much polish so I could get all my questions answered and then keep editing. I should have put a comment at the top mentioning that, my bad.